flag_groups.go

```diff
@@ -0,0 +1,144 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cobra
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	flag "github.com/spf13/pflag"
+)
+
+const (
+	RequiredAsGroup   = "cobra_annotation_required_if_others_set"
+	MutuallyExclusive = "cobra_annotation_mutually_exclusive"
+)
+
+func (c *Command) MarkFlagsRequiredTogether(flagNames ...string) {
+	for _, v := range flagNames {
+		f := c.Flags().Lookup(v)
+		if f.Annotations == nil {
+			f.Annotations = map[string][]string{}
+		}
+		// Each time this is called is a single new entry; this allows it to be a member of multiple groups if needed.
+		f.Annotations[RequiredAsGroup] = append(f.Annotations[RequiredAsGroup], strings.Join(flagNames, " "))
+	}
+}
+
+func (c *Command) MarkFlagsMutuallyExclusive(flagNames ...string) {
+	for _, v := range flagNames {
+		f := c.Flags().Lookup(v)
+		if f.Annotations == nil {
+			f.Annotations = map[string][]string{}
+		}
+		// Each time this is called is a single new entry; this allows it to be a member of multiple groups if needed.
+		f.Annotations[MutuallyExclusive] = append(f.Annotations[MutuallyExclusive], strings.Join(flagNames, " "))
+	}
+}
+
+// validateFlagGroups validates the mutuallyExclusive/requiredAsGroup logic.
+func (c *Command) validateFlagGroups() error {
+	if c.DisableFlagParsing {
+		return nil
+	}
+
+	flags := c.Flags()
+
+	// groupStatus format is the list of flags as a unique ID,
+	// then a map of each flag name and whether it is set or not.
+	groupStatus := map[string]map[string]bool{}
+	mutuallyExclusiveGroupStatus := map[string]map[string]bool{}
+	flags.VisitAll(func(pflag *flag.Flag) {
+		processFlagForGroupAnnotation(pflag, RequiredAsGroup, groupStatus)
+		processFlagForGroupAnnotation(pflag, MutuallyExclusive, mutuallyExclusiveGroupStatus)
+	})
+
+	errs := validateRequiredFlagGroups(groupStatus)
+	errsExclusive := validateExclusiveFlagGroups(mutuallyExclusiveGroupStatus)
+	errs = append(errs, errsExclusive...)
+
+	if len(errs) > 0 {
+		return combineErrors(errs, `, `)
+	}
+	return nil
+}
+
+func processFlagForGroupAnnotation(pflag *flag.Flag, annotation string, groupStatus map[string]map[string]bool) {
+	groupInfo, found := pflag.Annotations[annotation]
+	if found {
+		for _, group := range groupInfo {
+			if groupStatus[group] == nil {
+				groupStatus[group] = map[string]bool{}
+
+				flagnames := strings.Split(group, " ")
+				for _, name := range flagnames {
+					groupStatus[group][name] = false
+				}
+			}
+
+			groupStatus[group][pflag.Name] = pflag.Changed
+		}
+	}
+}
+
+func validateRequiredFlagGroups(data map[string]map[string]bool) []error {
+	var errs []error
+	for flagList, flagnameAndStatus := range data {
+		unset := []string{}
+		for flagname, isSet := range flagnameAndStatus {
+			if !isSet {
+				unset = append(unset, flagname)
+			}
+		}
+		if len(unset) == len(flagnameAndStatus) || len(unset) == 0 {
+			continue
+		}
+		sort.Strings(unset)
```

Comment:
Don't be shy to put that in a comment if you feel it is worth explaining.

Reply:
Thanks for the suggestion. I'm typically trying to get people to stop commenting unnecessarily but I guess I did feel it worth mentioning since you don't always sort those sorts of things.